### PR TITLE
[ntuple] Provide a page allocator that reuses previous memory allocations

### PR DESCRIPTION
**Pull request by jalopezg-git**
This pull request provides a page allocator that caches previous allocations up to a certain limit.  The cache is thread-local and defaults to 16 pages per thread (~1MiB per thread for default-sized pages).  This PR is largely based on the concepts in #8634.

Reuse of memory allocations in `RPageAllocatorCache` not only reduces calls to the global allocator but also heap fragmentation.  The global allocator is thread-safe albeit locked; thus, to also reduce contention, the cache in `RPageAllocatorCache` is thread-local.
Given the small size of the (per-thread) cache, the internal structure is a simple `std::deque`.  Anything more complex is not justified at the moment of this writing. 

**NOTE:** Performance evaluation (specially for the multi-threaded case) should be inserted here tomorrow.

## Changes or fixes:
- Add `RPageAllocatorCache`: this templated allocator returns pages that have at least the required capacity. `DeletePage()` does not immediately deallocate memory; instead, pages are returned to a thread-local cache, dropping the smallest allocated buffer if the cache is full. If a previous page cannot be recycled, the underlying allocator `AllocT` (`RPageAllocatorHeap` by default) is used to allocate memory.
- Change the signature of `RPageSource::UnsealPage()`: `SealPage()`, the counterpart of `UnsealPage()` takes an RPage and returns an RSealedPage.  Make the interface of `UnsealPage()` symmetric and return an RPage.
- Make `UnsealPage()` use `RPageAllocatorCache` by default.
- Remove stray `RPageAllocator{File,Daos}` classes.

## Checklist:
- [X] tested changes locally
- [X] updated the docs (if necessary)

This PR supersedes #8634.

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RPageAllocator.hxx [hunk 1]
```diff
@@ -77,6 +79,56 @@ public:
...
+This allocator returns pages that have at least the required capacity. `DeletePage()` does not immediately deallocate
+memory; instead, pages are returned to a thread-local cache, dropping the smallest allocated buffer if the cache is full.
+If a previous page cannot be recycled, `AllocT` is used to allocate memory.
+*/
+// clang-format on
+template <typename AllocT = RPageAllocatorHeap, std::size_t CacheSize = 16>
+class RPageAllocatorCache {
+public:
+   thread_local static struct RPageCache {
+      std::deque<RPage> fPages;
```
Comment (pcanal) [inline tree/ntuple/v7/inc/ROOT/RPageAllocator.hxx:97]: Does `fPages` have to be public?

Comment (jalopezg-git) [reply tree/ntuple/v7/inc/ROOT/RPageAllocator.hxx:96]: Misplaced `public:`; thanks for spotting!

**Review by jblomer — CHANGES_REQUESTED** [review]
I think we need to discuss it a little more:
- I don't think a thread-local page cache works. In `UnzipClusterImpl()` we allocate pages in (TBB) tasks.
- I think we should pass a pre-allocated page to `UnsealPage()` instead of the column id. That would result in symmetric appearance of the cache page allocator for allocation an deallocation.

**Comment by jalopezg-git** [comment]
> The cache mechanics looks good, but I think we need to discuss it a little more:
> 
>     * I don't think a thread-local page cache works. In `UnzipClusterImpl()` we allocate pages in (TBB) tasks.
> 

Oh, you are absolutely right!  I didn't thought enough about the implications of this; it might actually happen that pages are returned by a totally different thread (not TBB's).  Simplest thing would be to go for a locked data structure (while trying to keep the critical section small).

>     * I think we should pass a pre-allocated page to `UnsealPage()` instead of the column id. That would result in symmetric appearance of the cache page allocator for allocation an deallocation.

I agree that it would result in symmetric use; however, `UnsealPage()` has to reserve an additional buffer for unpacking the page, if needed.  Thus, I would leave it as-is or pass a page allocator as an additional (maybe template) argument.

**Pull request closed without merging**